storage/oasis/nodeapi/enigma/convert.go

```diff
@@ -0,0 +1,480 @@
+package enigma
+
+import (
+	"strings"
+
+	"github.com/oasisprotocol/oasis-core/go/common/cbor"
+	"github.com/oasisprotocol/oasis-core/go/common/crypto/signature"
+	"github.com/oasisprotocol/oasis-core/go/common/quantity"
+
+	// nexus-internal data types.
+	coreCommon "github.com/oasisprotocol/oasis-core/go/common"
+
+	"github.com/oasisprotocol/nexus/coreapi/v22.2.11/common/node"
+	genesis "github.com/oasisprotocol/nexus/coreapi/v22.2.11/genesis/api"
+	governance "github.com/oasisprotocol/nexus/coreapi/v22.2.11/governance/api"
+	registry "github.com/oasisprotocol/nexus/coreapi/v22.2.11/registry/api"
+	scheduler "github.com/oasisprotocol/nexus/coreapi/v22.2.11/scheduler/api"
+	staking "github.com/oasisprotocol/nexus/coreapi/v22.2.11/staking/api"
+	upgrade "github.com/oasisprotocol/nexus/coreapi/v22.2.11/upgrade/api"
+
+	apiTypes "github.com/oasisprotocol/nexus/api/v1/types"
+	"github.com/oasisprotocol/nexus/common"
+	"github.com/oasisprotocol/nexus/storage/oasis/nodeapi"
+
+	// data types for Enigma gRPC APIs.
+
+	nodeEnigma "github.com/oasisprotocol/nexus/coreapi/v23.0/common/node"
+	txResultsEnigma "github.com/oasisprotocol/nexus/coreapi/v23.0/consensus/api/transaction/results"
+	genesisEnigma "github.com/oasisprotocol/nexus/coreapi/v23.0/genesis/api"
+	governanceEnigma "github.com/oasisprotocol/nexus/coreapi/v23.0/governance/api"
+	registryEnigma "github.com/oasisprotocol/nexus/coreapi/v23.0/registry/api"
+	schedulerEnigma "github.com/oasisprotocol/nexus/coreapi/v23.0/scheduler/api"
+	stakingEnigma "github.com/oasisprotocol/nexus/coreapi/v23.0/staking/api"
+)
+
+func convertProposal(p *governanceEnigma.Proposal) *governance.Proposal {
```

Comment:
ok so how come we're converting from e* to damask instead of always forward?

Reply:
The PR description of #548 alludes to it:
> we (ab)use the v22.x types for two purposes:
> 1) For serde of gRPC requests/responses (as we also do for Cobalt, and later E)
> 2) As nexus's internal data representation (this needs a few more enums to interpret/manipulate; in an ideal world, we'd create our own types here)

So philosophically in this file, we're converting to `nodeapi` types, i.e. types that are internal to nexus. They are currently almost entirely a subset of damask types, but they can and will gain fields to be able to represent relevant data from all versions of oasis-core past and future.

Reply:
ah